lib/stringifier.js

```diff
@@ -1,8 +1,70 @@
 'use strict';
 
+/**
+ * @typedef {import('./types').XastParent} XastParent
+ * @typedef {import('./types').XastRoot} XastRoot
+ * @typedef {import('./types').XastElement} XastElement
+ * @typedef {import('./types').XastInstruction} XastInstruction
+ * @typedef {import('./types').XastDoctype} XastDoctype
+ * @typedef {import('./types').XastText} XastText
+ * @typedef {import('./types').XastCdata} XastCdata
+ * @typedef {import('./types').XastComment} XastComment
+ * @typedef {import('./types').StringifyOptions} StringifyOptions
+ */
+
 const { textElems } = require('../plugins/_collections.js');
 
-var defaults = {
+/**
+ * @typedef {{
+ *   width: void | string,
+ *   height: void | string,
+ *   textContext: null | XastElement,
+ *   indentLevel: number,
+ * }} State
+ */
+
+/**
+ * @typedef {{
+ *   doctypeStart: string,
+ *   doctypeEnd: string,
+ *   procInstStart: string,
+ *   procInstEnd: string,
+ *   tagOpenStart: string,
+ *   tagOpenEnd: string,
+ *   tagCloseStart: string,
+ *   tagCloseEnd: string,
+ *   tagShortStart: string,
+ *   tagShortEnd: string,
+ *   attrStart: string,
+ *   attrEnd: string,
+ *   commentStart: string,
+ *   commentEnd: string,
+ *   cdataStart: string,
+ *   cdataEnd: string,
+ *   textStart: string,
+ *   textEnd: string,
+ *   indent: string,
+ *   regEntities: RegExp,
+ *   regValEntities: RegExp,
+ *   encodeEntity: (char: string) => string,
+ *   pretty: boolean,
+ *   useShortTags: boolean
+ *   eol: 'lf' | 'crlf',
+ *   finalNewline: boolean,
+ * }} Options
```

Comment:
Can't we replace that with `@typedef {Required<StringifyOptions>} Options`?

Reply:
indent types are different but maybe easier to store it in state instead